src/fastify-plugins/maps/index.js

```diff
@@ -20,57 +32,158 @@ export const plugin = fp(mapsPlugin, {
+async function routes(fastify, opts) {
+  const { defaultOnlineStyleUrl = DEFAULT_MAPBOX_STYLE_URL } = opts
+
+  fastify.get(
+    '/style.json',
+    { schema: { querystring: GetStyleJsonQueryStringSchema } },
+    async (req, rep) => {
+      const serverAddress = await getFastifyServerAddress(req.server.server)
+
+      // 1. Attempt to get "default" local static map's style.json
+      {
+        const styleId = 'default'
+
+        const results = await Promise.all([
+          fastify.mapeoStaticMaps.getStyleJsonStats(styleId),
+          fastify.mapeoStaticMaps.getResolvedStyleJson(styleId, serverAddress),
+        ]).catch(() => {
+          fastify.log.warn('Cannot read default static map')
+          return null
+        })
+
+        if (results) {
+          const [stats, styleJson] = results
+          rep.headers(createStyleJsonResponseHeaders(stats.mtime))
+          return styleJson
+        }
+      }
+
+      // 2. Attempt to get a default style.json from online source
+      {
+        const { key } = req.query
+
+        const upstreamUrlObj = new URL(defaultOnlineStyleUrl)
+        const { hostname } = upstreamUrlObj
+
+        if (key) {
+          if (isRecognizedMapProvider(hostname)) {
+            const paramToUpsert = getMapProviderApiKeyQueryParamName(hostname)
+
+            // Note that even if the search param of interest already exists in the url
+            // it is overwritten by the key provided in the request's search params
+            upstreamUrlObj.searchParams.set(paramToUpsert, key)
+          } else {
+            fastify.log.warn(
+              `Provided API key will not be applied to unrecognized provider: ${hostname}`
+            )
+          }
+        }
+
+        try {
+          const upstreamResponse = await fetch(upstreamUrlObj.href, {
+            signal: AbortSignal.timeout(3000),
+          })
+
+          if (upstreamResponse.ok) {
+            // Set up headers to forward
+            // TODO: Change this to an allow-list of headers instead of a block-list
+            for (const [name, value] of upstreamResponse.headers) {
+              // We do our own content encoding
+              if (name.toLowerCase() === 'content-encoding') continue
+              rep.header(name, value)
+            }
+            // Some upstream providers will not set the 'application/json' content-type header despite the body being JSON e.g. Protomaps
+            // TODO: Should we forward the upstream 'content-type' header?
+            // We kind of assume that a Style Spec-compatible JSON payload will always be used by a provider
+            // Technically, there could be cases where a provider doesn't use the Mapbox Style Spec and has their own format,
+            // which may be delivered as some other content type
+            rep.header('content-type', 'application/json; charset=utf-8')
+            return upstreamResponse.json()
+          } else {
+            fastify.log.error(
```

Comment:
Short answer: yes, I think it should be a warning.

---

Long answer:

At a high level, I believe two things about how errors should be logged:

1. _Errors are for developer mistakes._ Invalid states are a good example, because a perfect developer would never put the system into an invalid state.
1. _Warnings are for unavoidable problems._ Network failures are usually a good example, because even a perfect developer can't fix a user turning their wifi off.

To be clear, this is personal preference!

So what would I do this case?

If we get a 5xx error from the server, that's a _warning_, because we didn't make a mistake. But you could argue that we should _error_ if we get a 4xx. After all, if we send a bogus request, isn't that a developer's fault?

In this case, I think the answer is..._maybe_? Do we consider a bad API key an error, or a warning? Does it depend on what kind of 4xx status code you get (e.g., is 422 "our fault" but 429 is not)? I don't think it's worth thinking too hard about, so I'd probably just go with a warning.

Reply:
this is helpful! thanks for sharing both answers 😄

Reply:
addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/0d992efff1f0e0f13b17a89478bd4dbd32079fed